medicaltorch/datasets.py

```diff
@@ -214,26 +252,36 @@ def __init__(self, filename_pairs, slice_axis=2, cache=True,
         self._prepare_indexes()
 
     def _load_filenames(self):
-        for input_filename, gt_filename in self.filename_pairs:
-            segpair = SegmentationPair2D(input_filename, gt_filename,
-                                         self.cache, self.canonical)
-            self.handlers.append(segpair)
+        for input_filename, gt_filename, roi_filename, metadata, modality in self.filename_pairs:
+            segpair = SegmentationPair2D(input_filename, gt_filename, metadata=metadata,
+                                         cache=self.cache, canonical=self.canonical)
+            roipair = SegmentationPair2D(input_filename, roi_filename, metadata=metadata,
+                                         cache=self.cache, canonical=self.canonical)
 
-    def _prepare_indexes(self):
-        for segpair in self.handlers:
-            input_data_shape, _ = segpair.get_pair_shapes()
-            for segpair_slice in range(input_data_shape[2]):
+            self.handlers.append([segpair, roipair])
 
-                # Check if slice pair should be used or not
-                if self.slice_filter_fn:
-                    slice_pair = segpair.get_pair_slice(segpair_slice,
-                                                        self.slice_axis)
+    def _prepare_indexes(self):
+        for seg_roi_pairs in self.handlers:
+            seg_pair, roi_pair = seg_roi_pairs
+            input_data_shape, _ = seg_pair.get_pair_shapes()
+
+            for idx_pair_slice in range(input_data_shape[self.slice_axis]):
+                # if ROI provided, filter empty (img, roi) slices
+                if self.slice_filter_fn and slice_roi_pair['gt'] is not None:
+                    slice_roi_pair = roi_pair.get_pair_slice(idx_pair_slice,
+                                                                self.slice_axis)
+                    filter_fn_ret_roi = self.slice_filter_fn(slice_roi_pair)
```

Comment:
for multichannel I put a for loop to pass a single input at the time (https://github.com/neuropoly/medicaltorch/pull/1/commits/4bc295917f12f1d08a38af72fbf372b85f9d7bbb#diff-62669d245379d6c761170c1aee0cda8bR270). Without it should crash since the input file is now a list

Reply:
Great! Will implement something similar, will be back soon

Reply:
Done